lib/matplotlib/figure.py

```diff
@@ -1242,6 +1242,12 @@ def colorbar(
             fig.sca(current_ax)
             cax.grid(visible=False, which='both', axis='both')
 
+        if hasattr(mappable, "figure") and mappable.figure is not self.figure:
+            _log.warning(
```

Comment:
Did we want a _log.warning or a warning.warning?  How bad do we think it is that a user has put a colorbar on a different figure?

Reply:
We should probably use `warn_external` rather that `_log.warning`.

I think we should also only add this warning if we inferred what axes to steal from 

```
import matplotlib.pyplot as plt
fig, ax = plt.subplots()
fig2, (axA, axB) = plt.subplots(2)
im = ax.imshow([[1, 2], [3, 4]])

fig2.colorbar(im)   # this should warn

fig2.colorbar(im, ax=ax)   # this is weird, but should not warn
fig2.colorbar(im, ax=axA)  # this should not warn
fig2.colorbar(im, cax=axB) # also should not warn
```

Reply:
Yes, `warn_external` is the way to go.

> I think we should also only add this warning if we inferred what axes to steal from

I'm not convinced. Generally, it is dubious if the mappable is in another figure than the colorbar, because there's no guarantee that both are shown together. Scenarios I could imagine are somewhat contrieved: (1) you have two figures with images in a paper, but only want to show one colorbar; (2) You have a GUI app and want to render the colorbar in a separate widget. In such cases I would recommend to reconsider whether that's actually a good idea (which the warning would serve to). And if somebody is convinced this is still the right path to take, they can filter the warning.

In turn, it's quite likely that the user has mixed up some fig/ax variables when mappable and colorbar are on different figures, e.g. `fig2.colorbar(im, ax=ax)` above could very well have been a typo and `fig.colorbar(im, ax=ax)` was actually intended.

Reply:
I agree with @timhoffm. I think we should warn anyway, and users can decide if it is a mistake or what they intend to do. For now I'm going to go ahead with this and submit a new commit which also includes some code style fixes and hopefully we can resolve this conversation.